### PR TITLE
feature: supported async SSL session fetch callback with OpenSSL 1.1.1.

### PR DESCRIPTION
**Pull request by spacewander**
Fix openresty/openresty#456.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

**Comment by spacewander** [comment]
Depended on: https://github.com/openresty/openresty/pull/474

File: src/ngx_http_lua_module.c [hunk 1]
```diff
@@ -1077,6 +1077,14 @@ ngx_http_lua_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
 
             return NGX_CONF_ERROR;
 #else
+
+#   ifdef SSL_ERROR_WANT_CLIENT_HELLO_CB
+
```
Comment (thibaultcha) [inline src/ngx_http_lua_module.c]: style: extra line intended here?

File: .travis.yml [hunk 2]
```diff
@@ -50,8 +50,7 @@ env:
   matrix:
     - NGINX_VERSION=1.15.8 OPENSSL_VER=1.0.2q OPENSSL_OPT="" OPENSSL_PATCH_VER=1.0.2h
     - NGINX_VERSION=1.15.8 OPENSSL_VER=1.1.0j OPENSSL_OPT="" OPENSSL_PATCH_VER=1.1.0d
-    # TODO: when adding an OpenSSL version >= 1.1.1, please add "enable-tls1_3"
-    # to $OPENSSL_OPT.
+    - NGINX_VERSION=1.15.8 OPENSSL_VER=1.1.1b OPENSSL_OPT="enable-tls1_3" OPENSSL_PATCH_VER=""
```
Comment (thibaultcha) [inline .travis.yml]: - Could we bump to 1.1.1c?
- I saw the TODO note on this diff, but nonetheless, TLS 1.3 is enabled by default in OpenSSL 1.1.1 afaik. Do we really need the `enable-tls1_3` option?

File: src/ngx_http_lua_ssl_session_fetchby.h [hunk 3]
```diff
@@ -32,6 +32,10 @@ ngx_ssl_session_t *ngx_http_lua_ssl_sess_fetch_handler(
     u_char *id, int len, int *copy);
 #endif
 
+#if HAVE_SSL_CLIENT_HELLO_CB_SUPPORT
+int ngx_http_lua_ssl_client_hello_handler(ngx_ssl_conn_t *ssl_conn,
+    int *al, void *arg);
+#endif
```
Comment (thibaultcha) [inline src/ngx_http_lua_ssl_session_fetchby.h]: This declaration should be within the above `#if (NGX_HTTP_SSL)` branch.

File: src/ngx_http_lua_ssl_session_fetchby.c [hunk 4]
```diff
@@ -376,10 +361,110 @@ ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
...
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+
+#ifdef SSL_ERROR_PENDING_SESSION
+
+        return SSL_magic_pending_session_ptr();
+
+#else
+
```
Comment (thibaultcha) [inline src/ngx_http_lua_ssl_session_fetchby.c]: style: do we need an extra line here? Many precedents in the codebase suggest we don't.

Comment (spacewander) [reply src/ngx_http_lua_ssl_session_fetchby.c]: @thibaultcha 
But `#ifdef SSL_ERROR_PENDING_SESSION` adds an extra blank line below?

File: src/ngx_http_lua_ssl_session_fetchby.h (hunk 3 above)
Comment (thibaultcha) [reply src/ngx_http_lua_ssl_session_fetchby.h]: @spacewander Ping

**Comment by thibaultcha** [comment]
@spacewander We also need to update the README which mentions:

> If you are using OpenSSL libraries not provided by OpenResty, then you need to apply the following patch for OpenSSL 1.0.2h or later:

To mention that no such patch is necessary anymore starting with OpenSSL 1.1.1.

**Comment by spacewander** [comment]
@thibaultcha 
Updated.

**Comment by agentzh** [comment]
@spacewander I'm getting the following compiler error while building this branch with OpenSSL 1.1.1c:

```
/home/agentzh/git/lua-nginx-module/src/ngx_http_lua_ssl_session_fetchby.c: In function ‘ngx_http_lua_ssl_sess_fetch_handler’:
/home/agentzh/git/lua-nginx-module/src/ngx_http_lua_ssl_session_fetchby.c:446:38: error: ‘c’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         ngx_log_error(NGX_LOG_CRIT, c->log, 0,
                                      ^~
```

**Comment by agentzh** [comment]
@spacewander Same compiler error as above with OpenSSL 1.1.0k.

**Comment by spacewander** [comment]
@agentzh 
Fixed in the new commit.

File: src/ngx_http_lua_ssl_session_fetchby.c [hunk 5]
```diff
@@ -376,10 +361,124 @@ ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
...
+    c = ngx_ssl_get_connection(ssl_conn);
+
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+
+#ifndef SSL_ERROR_PENDING_SESSION
+
+        return SSL_magic_pending_session_ptr();
```
Comment (thibaultcha) [inline src/ngx_http_lua_ssl_session_fetchby.c]: When compiling this patch _without_ the nginx `ssl_sess_cb_yield` patch, one now gets this warning:
```
gcc -c -I/home/chasum/code/openresty/openresty-dev/ngx_lua/deps/luajit2-v2.1-agentzh/include//luajit-2.1  -pipe  -O -W -Wall -Wpointer-arith -Wno-unused-parameter -Werror -g -O0 -g -DNGXs/include -I/home/chasum/code/openresty/openresty-dev/ngx_lua/deps/include -DNDK_SET_VAR -DNDhome/chasum/code/openresty/openresty-dev/ngx_lua/ngx_devel_kit/objs -I objs/addon/ndk -I /homresty/openresty-dev/ngx_lua/deps/include/libdrizzle-1.0 -I /home/chasum/code/openresty/openreresty-dev/ngx_lua/stream-lua-nginx-module/src/api -I objs -I src/http -I src/http/modules -I um/code/openresty/openresty-dev/ngx_lua/ngx_devel_kit/src -I /home/chasum/code/openresty/opendev/ngx_lua/deps/luajit2-v2.1-agentzh/include/luajit-2.1 -I src/stream \
        -o objs/addon/src/ngx_http_headers_more_util.o \
        /home/chasum/code/openresty/openresty-dev/ngx_lua/headers-more-nginx-module/src/ngx_h
/home/chasum/code/openresty/openresty-dev/ngx_lua/lua-nginx-module/src/ngx_http_lua_ssl_sessi
/home/chasum/code/openresty/openresty-dev/ngx_lua/lua-nginx-module/src/ngx_http_lua_ssl_sessid you mean ‘SSL_has_matching_session_id’? [-Werror=implicit-function-declaration]
         return SSL_magic_pending_session_ptr();
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                SSL_has_matching_session_id
/home/chasum/code/openresty/openresty-dev/ngx_lua/lua-nginx-module/src/ngx_http_lua_ssl_sessia ‘struct ssl_session_st *’} makes pointer from integer without a cast [-Werror=int-conversio
         return SSL_magic_pending_session_ptr();
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

Seems like this branch needs to be better guarded.

File: src/ngx_http_lua_ssl_session_fetchby.c [hunk 6]
```diff
@@ -376,10 +361,125 @@ ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
...
+
+    *copy = 0;
+
+    c = ngx_ssl_get_connection(ssl_conn);
+
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+
+#ifndef SSL_ERROR_PENDING_SESSION
```
Comment (thibaultcha) [inline src/ngx_http_lua_ssl_session_fetchby.c]: It seems to me that this branch should remain `#ifdef SSL_ERROR_PENDING_SESSION`.

**Comment by thibaultcha** [comment]
I just added another commit to mention the nginx core patches for this feature that we failed to document for 1.13.6 cores and above.

@agentzh I'd suggest you have a look at the wording etc. It seems that we combined both patches for nginx 1.11.2 (http://openresty.org/download/nginx-1.11.2-nonblocking_ssl_handshake_hooks.patch) but we haven't done so for the 1.13.6 and 1.15.8 cores.

File: src/ngx_http_lua_ssl_session_fetchby.c (hunk 6 above)
Comment (spacewander) [reply src/ngx_http_lua_ssl_session_fetchby.c]: :cry: My bad! I should not submit this debug tweak.

**Comment by thibaultcha** [comment]
The latest commit I pushed follows my above suggestion on the `#ifdef SSL_ERROR_PENDING_SESSION` branch and seems to resolve the Travis CI failures so far.

**Comment by agentzh** [comment]
@thibaultcha Yeah, we use separate nginx core patches as documented here:

https://openresty.org/en/nginx-ssl-patches.html

File: README.markdown [hunk 7]
```diff
@@ -2621,16 +2621,45 @@ But do not forget to comment this line out before publishing your site to the wo
...
+
+For the 1.15.8 NGINX core:
+
+<https://github.com/openresty/openresty/blob/master/patches/nginx-1.15.8-ssl_cert_cb_yield.patch>
+<https://github.com/openresty/openresty/blob/master/patches/nginx-1.15.8-ssl_pending_session.patch>
+
+For subsequent NGINX cores, the patches will be updated and their name will
+contain the supported core version (e.g.
+`nginx-<major.minor.patch>-*.patch`).
+
```
Comment (agentzh) [inline README.markdown]: Let's just use the openresty.org links here:

https://openresty.org/en/nginx-ssl-patches.html

https://openresty.org/en/openssl-patches.html

It's too verbose to put them here in this lengthy README.

Comment (thibaultcha) [reply README.markdown]: @agentzh Updated and now much more concise with better wording. We also take the opportunity to inform users of our OpenSSL packages (https://openresty.org/en/linux-packages.html):

> If you are not using one of the OpenSSL packages provided by OpenResty, you will need to apply patches to OpenSSL 1.0.2, up to (and including) 1.1.0:
>
> https://openresty.org/en/openssl-patches.html
>
> Similarly, if you are not using the NGINX core shipped with OpenResty 1.11.2.1 or later, you will need to apply patches to the standard NGINX core:
>
> https://openresty.org/en/nginx-ssl-patches.html

**Comment by thibaultcha** [comment]
Very happy to see this in, great job @spacewander.

**Pull request merged**